storage/index.go

```diff
@@ -265,16 +264,57 @@ func (i *nsIndex) Query(
 		return index.QueryResults{}, err
 	}
 
-	idxIter := index.NewIterator(i.nsID, iter, i.opts)
+	var (
+		brokeEarly = false
+		numAdded   = 0
+		results    = i.opts.ResultsPool().Get()
+		cleanupFn  = func() {
```

Comment:
nit: Just a thought, you can do what I do sometimes and:
```
success := false
defer func() {
  if !success {
    // .. do cleanup
  }
}()

// do work

success = true
return ...
```

Reply:
Nice, yeah that's a lot better.

Reply:
done